modules/data.atmosphere/R/met_temporal_downscale.Gaussian_ensemble.R

```diff
@@ -0,0 +1,398 @@
+  # Reading in the source data
+  source <- list()
+  tem <- ncdf4::nc_open(input_met)
+  for (j in seq_along(var$CF.name)) {
+    if (exists(as.character(var$CF.name[j]), tem$var) == FALSE) {
+      source[[j]] <- NA
+    } else {
+      source[[j]] <- ncdf4::ncvar_get(tem, as.character(var$CF.name[j]))
+    }
+  }
+  ncdf4::nc_close(tem)
+  
+  source <- data.frame(source)
+  colnames(source) <- var$CF.name
+  
+  # Default downscale will be to the resolution of the training dataset
+  reso <- 24/(nrow(train)/365)
+  reso_len <- nrow(train)
+  
+  # If a source doesn't have leap days, we need to remove them to ensure equal
+  # lengths
+  if (lubridate::leap_year(year) == TRUE) 
+  {
+    if (length(source$air_temperature)%%366 > 0) {
+      if (length(train$air_temperature)%%366 == 0) {
+        train <- train[1:365 * (nrow(train)/366)]
+      }
+      eph_year <- 2006
+    }
+  }  #chose a non-leap year to use for daylength calculations if we don't have the 
+  if (lubridate::leap_year(eph_year) == TRUE) {
+    sp <- 366
+  }
+  if (lubridate::leap_year(eph_year) == FALSE) {
+    sp <- 365
+  }
+  
+  # Now we start a for loop for the ensemble members and begin downscaling. A
+  # random normal distribution is used to downscale as so;
+  # (mean <- value of source data) (sd <- +/- window_days of train data at the
+  # same time intervals) 
+  for (e in seq_along(1:ensemble_members)) {
+    
+    div <- nrow(train)/nrow(source)  #tells us how many values need to be generated (via downscaling) from each source value
+    sd_step <- nrow(train)/sp  #allows us to step through each window at specific times
+    
+    # Temperature For temperature, we create a random normal distribution with the
+    # mean of the source value and the standard deviation of the training value
+    # within the window, and step through that window sequentially so that we are
+    # only using the standard devaition of values recorded at the same time as the
+    # value we are generating.
+    temp <- c("air_temperature", "air_temperature_max", "air_temperature_min")
+    df <- data.frame()
+    for (u in temp) {
+      woo <- vector()
+      a <- as.numeric(train[[u]])
+      sour <- as.numeric(source[[u]])
+      if (all(is.na(sour)) == TRUE) {
+        woo <- rep(NA, reso_len)
+      }
+      if (all(is.na(sour)) == FALSE) {
+        for (x in seq_along(sour)) {
+          four <- vector()
+          for (n in seq_along(1:div)) {
+            lowday <- x * div - wd * sd_step + n
+            highday <- x * div + wd * sd_step + n
+            if (lowday < 0) {
+              lowday <- x + n
+              highday <- x + n + wd * sd_step + n
+            }
+            if (highday > reso_len) {
+              highday <- reso_len
+              lowday <- x * div - wd * sd_step + n
+            }
+            four[n] <- rnorm(1, mean = sour[x], sd = sd(a[seq(from = lowday, 
+                                                              to = highday, by = sd_step)]))
```

Comment:
Ok, sounds good. Yeah I've been trying to keep up with Christy's code the best I can but communicating has been rather difficult (normally takes her at least a week to respond to an email) so I kinda had to push forth alone. But this feedback is good, thank you for the explanations. Just to make sure I'm understanding this correctly, let's say we want to downscale a MACA product (daily res) at Harvard Forest. We have ~ 10 years of overlapping data, so we use linear regression to debias the MACA data and that's the 'training' phase. So we use the same regression in the 'application' phase to debias MACA data in all years without overlapping data. At what point do we downscale from daily to half hourly?